lib/RunTests.js

```diff
@@ -29,21 +28,55 @@ function getPipeFilename(runsExecuted /*: number */) /*: string */ {
     : `/tmp/elm_test-${process.pid}.sock`;
 }
 
-function infoLog(
-  report /*: typeof Report.Report */,
-  msg /*: string */
-) /*: void */ {
-  if (report === 'console') {
-    console.log(msg);
-  }
-}
-
-function clearConsole(report /*: typeof Report.Report */) {
-  if (report === 'console') {
-    process.stdout.write(
-      process.platform === 'win32' ? '\x1B[2J\x1B[0f' : '\x1B[2J\x1B[3J\x1B[H'
-    );
-  }
+// This could have been a class, but I couldn’t figure out how to type it with
+// Flow comments and Prettier.
+// This lets you log something like `elm.json changed > Compiling > Starting tests`
+// where each segment appears over time.
+// `\r` moves the cursor to the start of the line. This is important when there
+// are Elm compilation errors - they overwrite the progress text rather than
+// weirdly ending up with stuff like:
+// `elm.json changed > Compiling-- NAMING ERROR - File.elm`
+// Also note that using too much `clearLine` or `clearConsole` causes flickering
+// on Windows, so it's nicer to cleverly overwrite old output when possible.
+function makeProgressLogger(report /*: typeof Report.Report */) {
+  const items = [];
+  return {
+    log(message) {
+      items.push(message);
+      if (!Report.isMachineReadable(report)) {
+        process.stdout.write(`${items.join(' > ')}\r`);
+      }
+    },
+    overwrite(message) {
+      items.length = 0;
+      items.push(message);
+      if (!Report.isMachineReadable(report)) {
+        process.stdout.write(`${message}\r`);
+      }
+    },
+    clearLine() {
+      items.length = 0;
+      if (!Report.isMachineReadable(report)) {
+        readline.clearLine(process.stdout, 0);
+      }
+    },
+    clearConsole() {
+      items.length = 0;
+      if (!Report.isMachineReadable(report)) {
+        process.stdout.write(
+          process.platform === 'win32'
+            ? '\x1B[2J\x1B[0f'
+            : '\x1B[2J\x1B[3J\x1B[H'
+        );
+      }
+    },
+    newLine() {
```

Comment:
Could we replace this function with `logLine`? 

```js
logLine(message) {
  items.length = 0;
  if (!Report.isMachineReadable(report)) {
    process.stdout.write(message + `\n`);
  }
}
```

The `progressLog

Reply:
We need `progressLogger.newLine()` in the `catch` part though, so the error message appears _below_ the progress line (rather than (partly) on top of it.)

Reply:
We could use `processLogger.logLine("")` to print the empty line.